### PR TITLE
MSR: Create instances of NSObjects and INativeObjects without using reflection

### LINKED ISSUE #18358: Research improving finding the constructor to call to create a managed NSObject instance in MSR
Ref: last part of https://github.com/dotnet/runtime/issues/86649#issuecomment-1562617121

### When instantiating a managed NSObject instance (or subclass) for an existing Objective-C object

Say we have the following managed class:

```cs
class MyObject : NSObject {
    [Export ("doSomething:")]
    public static void DoSomething (NSString something) {}
}

class NSString : NSObject {
    protected NSString (IntPtr handle) : base (handle) {}
}
```

In this case we do something like:

```cs
partial class MyObject {
    class __RegistrarHelper_ {
        [UnmanagedCallersOnly ("__MyObject_DoSomething__")]
        static void DoSomething (IntPtr handle, IntPtr selector, IntPtr arg1)
        {
             // Runtime.GetNSObject:
             // https://github.com/xamarin/xamarin-macios/blob/3dd412daa6b5522029014b0be298f27d751c5f45/src/ObjCRuntime/Runtime.cs#L1516-L1519
             // basically something like this:
             //     if we already have a managed instance for the handle 'handle', return that
             //     otherwise find the constructor that takes an IntPtr (using reflection), and call that constructor
            var arg1Obj = Runtime.GetNSObject ();
                
            MyObject.DoSomething (arg1Obj);
        }
    }
}
```

This is in some ways very similar to Filip's INativeObject/Selector case from above, except a bit more complex, because `Runtime.GetNSObject` will look up the actual runtime type of the Objective-C `handle`, find the closest matching C# type, and create an instance of that type. This might very well be a subclass of the type the caller expects (and this is why we can't just do a `new NSString (handle)` in the generated code).

Potential solution: we already generate a table to map Objective-C class -> managed type (which we use to find the closest matching C# type for a given Objective-C classr), we could add the token for the IntPtr ctor to table, so that when we find a managed type, we'll find the ctor to call too (or maybe even generate code to call the ctor directly, because calling a method given a token in C# is rather obnoxious)

### PR DESCRIPTION
**Pull request by simonrozsival**
Closes #18358 

This PR adds lookup tables and factory methods for INativeObjects and NSObjects. These generated methods allow us to create instances of these objects without needing reflection.

### CODE REVIEW

File: src/Foundation/NSObject2.cs [hunk 1]
```diff
@@ -69,6 +69,15 @@ public class NSObjectFlag {
 	}
 #endif
 
+#if NET
+	public interface INSObjectFactory {
+		// The method will be implemented via custom linker step if the managed static registrar is used
+		// for NSObject subclasses which have an (NativeHandle) or (IntPtr) constructor.
+		[MethodImpl(MethodImplOptions.NoInlining)]
+		virtual static NSObject ConstructNSObject (NativeHandle handle) => null;
```
Comment (stephen-hawley) [inline src/Foundation/NSObject2.cs:77]: I'm always concerned about name collisions. Are we OK with injecting a new method into classes that we don't control and can't ensure that this method doesn't already exist? Will this be handled by the `NewSlot` flag? I've not used it before.

**Review by stephen-hawley — COMMENTED** [review]
Nifty idea, well done. I have some comments and some performance nits.

File: tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs [hunk 2]
```diff
@@ -281,6 +291,233 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
...
+				il.Emit (OpCodes.Newobj, ctor);
+				il.Emit (OpCodes.Ret);
+			} else {
+				throw new UnreachableException ();
+			}
+
+			Annotations.Mark (createInstanceMethod);
+		}
+
+		static MethodReference? FindNSObjectConstructor (TypeDefinition type)
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs:503]: Hey Steven. Thanks for the suggestion. You're right that it's not very efficient and it makes several unnecessary allocations. I'll update the implementation.

File: tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs [hunk 3]
```diff
@@ -473,13 +710,32 @@ static string GetMethodSignature (MethodDefinition method)
...
+			if (type.IsNested) {
+				if (!type.IsNestedPublic) {
+					type.IsNestedAssembly = true;
+				}
+
 				EnsureVisible (caller, type.DeclaringType);
-			} else {
-				type.IsPublic = true;
+			} else if (!type.IsPublic) {
+				type.IsNotPublic = true;
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs:738]: We need the type to be at least internal. Unless it's already public, we need to change its visibility attributes. There might be a better way to set the visibility modifier but this worked for me so far.

File: src/Foundation/NSObject2.cs (hunk 1 above)
Comment (simonrozsival) [reply src/Foundation/NSObject2.cs:77]: I haven't payed close attention to this problem so far. I've been trying to get it working first :) I definitely need to change something so that the new interfaces and methods aren't public and we don't introduce any new public APIs with this change.

I don't think this change can be implemented without adding the methods to the types that we want to create (especially because of generic types). We could in theory first check if there is a naming conflict and generate a different name if so. Or we could always add some random characters to decrease the likelihood that there will be a conflict.

File: src/ObjCRuntime/Runtime.cs [hunk 4]
```diff
@@ -1304,19 +1304,46 @@ static void MissingCtor (IntPtr ptr, IntPtr klass, Type type, MissingCtorResolut
...
 		// The 'selector' and 'method' arguments are only used in error messages.
-		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution) where T : class, INativeObject
+		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution) where T : NSObject
 		{
 			return ConstructNSObject<T> (ptr, type, missingCtorResolution, IntPtr.Zero, default (RuntimeMethodHandle));
 		}
 
-		// The generic argument T is only used to cast the return value.
 		// The 'selector' and 'method' arguments are only used in error messages.
-		static T? ConstructNSObject<T> (IntPtr ptr, Type type, MissingCtorResolution missingCtorResolution, IntPtr sel, RuntimeMethodHandle method_handle) where T : class, INativeObject
```
Comment (rolfbjarne) [inline src/ObjCRuntime/Runtime.cs:1316]: Why did you remove the `class` constraint? This method shouldn't be called with structs.

File: src/ObjCRuntime/Runtime.cs [hunk 5]
```diff
@@ -1304,19 +1304,46 @@ static void MissingCtor (IntPtr ptr, IntPtr klass, Type type, MissingCtorResolut
...
+				}
+
+				// If we couldn't create an instance of T through the factory method, we'll use the lookup table.
+				// This table can't instantiate generic types though.
+				if (!type.IsGenericType) {
+					instance ??= RegistrarHelper.ConstructNSObject<T> (type, nativeHandle);
+				}
+
+				if (instance is null) {
+					MissingCtor (ptr, IntPtr.Zero, type, missingCtorResolution, sel, method_handle);
```
Comment (rolfbjarne) [inline src/ObjCRuntime/Runtime.cs:1340]: We might need a different error message, because it would be rather confusing for customers if we said we couldn't find the constructor and it ended up actually being there.

File: src/ObjCRuntime/INativeObject.cs [hunk 6]
```diff
@@ -15,6 +16,13 @@ public interface INativeObject {
 			get;
 		}
 #endif
+
+#if NET
+		// The method will be implemented via custom linker step if the managed static registrar is used
+		// for classes which have an (NativeHandle, bool) or (IntPtr, bool) constructor.
+		[MethodImpl(MethodImplOptions.NoInlining)]
+		static virtual INativeObject? ConstructINativeObject (NativeHandle handle, bool owns) => null;
```
Comment (rolfbjarne) [inline src/ObjCRuntime/INativeObject.cs:24]: This is detected as a breaking change by our tooling.

I see two options:

1. Fix our api comparison tool that detects breaking changes to handle interfaces with default methods (and not consider them breaking).
2. Have the MSR code inject this method in the interface at build time.

File: tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs [hunk 7]
```diff
@@ -281,6 +291,233 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
...
+				il.Emit (OpCodes.And);
+				il.Emit (OpCodes.Brfalse_S, falseTarget);
+
+				il.Emit (OpCodes.Ldloc, instanceVariable);
+				il.Emit (OpCodes.Call, abr.Runtime_TryReleaseINativeObject);
+
+				il.Append (falseTarget);
+
+				il.Emit (OpCodes.Ldloc, instanceVariable);
+				il.Emit (OpCodes.Ret);
```
Comment (rolfbjarne) [inline tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs:481]: You don't actually need the variable:

You don't actually need the variable:

```suggestion
				il.Emit (OpCodes.Ldarg, nativeHandleParameter);
				if (nsobjectConstructor.Parameters [0].ParameterType.Is ("System", "IntPtr"))
					il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
				il.Emit (OpCodes.Newobj, nsobjectConstructor);

				var falseTarget = il.Create (OpCodes.Nop);
				il.Emit (OpCodes.Ldloc, instanceVariable);
				il.Emit (OpCodes.Ldnull);
				il.Emit (OpCodes.Cgt_Un);
				il.Emit (OpCodes.Ldarg, ownsParameter);
				il.Emit (OpCodes.And);
				il.Emit (OpCodes.Brfalse_S, falseTarget);

				il.Emit (OpCodes.Dup);
				il.Emit (OpCodes.Call, abr.Runtime_TryReleaseINativeObject);

				il.Append (falseTarget);

				il.Emit (OpCodes.Ret);
```

File: src/Foundation/NSObject2.cs (hunk 1 above)
Comment (rolfbjarne) [reply src/Foundation/NSObject2.cs:77]: > Or we could always add some random characters to decrease the likelihood that there will be a conflict.

Adding a `_Xamarin_` prefix should be enough I think

File: src/ObjCRuntime/Runtime.cs [hunk 8]
```diff
@@ -1368,6 +1455,25 @@ static void MissingCtor (IntPtr ptr, IntPtr klass, Type type, MissingCtorResolut
 			ctorArguments [1] = owns;
 
 			return (T?) ctor.Invoke (ctorArguments);
+
+#if NET
+			// This is a workaround for a compiler issue.
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs:1460]: I think the comment is outdated. I was running into an issue where the compiler would crash when `T.ConstructINativeObject` or `T.ConstructNSObject` was called in the parent method but it was probably caused by something else (or it was in combination with some invalid IL that I generated during earlier stages of development). I don't observe the same crash now that I rebuilt it with "direct" calls to those methods. I'll get rid of the comment and I think I'll also remove the local function.

[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs:1460]: Update: I've been able to find the issue that led me to move the code into a separate function. Without `T.ConstructINativeObject(...)` being called from different function than `Runtime.ConstructINativeObject<T>`, for example calling `CNContactFormatter.GetDescriptorForRequiredKeys (CNContactFormatterStyle.PhoneticFullName)` causes a crashes with:

```
2023-06-30 17:11:06.854 MySingleView[90723:11572593] error: * Assertion at /Users/runner/work/1/s/src/mono/mono/mini/mini-generic-sharing.c:2283, condition `m_class_get_vtable (info->klass)' not met

=================================================================
        Native Crash Reporting
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================

=================================================================
        Native stacktrace:
=================================================================
        0x10362265c - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_dump_native_crash_info
        0x1035e1578 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_handle_native_crash
        0x1038001a0 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : sigabrt_signal_handler.cold.1
        0x103621f10 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_runtime_setup_stat_profiler
        0x18cd26a24 - /usr/lib/system/libsystem_platform.dylib : _sigtramp
        0x18ccf7c28 - /usr/lib/system/libsystem_pthread.dylib : pthread_kill
        0x18cc05ae8 - /usr/lib/system/libsystem_c.dylib : abort
        0x102e39bf0 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libxamarin-dotnet.dylib : _ZL14print_callbackPKci
        0x10365e4b8 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : monoeg_g_logv_nofree
        0x10365e570 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : monoeg_assertion_message
        0x10365e5b4 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mono_assertion_message_unreachable
        0x1035eaa8c - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : instantiate_info
        0x1035e9900 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mini_instantiate_gshared_info
        0x103595248 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MonoBundle/libmonosgen-2.0.dylib : mini_init_method_rgctx
        0x1026c2acc - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MacOS/MySingleView : ObjCRuntime_Runtime_ConstructINativeObject_T_REF_intptr_bool_System_Type_ObjCRuntime_Runtime_MissingCtorResolution_intptr_System_RuntimeMethodHandle
        0x1026c5788 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MacOS/MySingleView : ObjCRuntime_Runtime_GetINativeObject_T_REF_intptr_bool_System_Type_bool_intptr_System_RuntimeMethodHandle
        0x1026c4860 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MacOS/MySingleView : ObjCRuntime_Runtime_GetINativeObject_T_REF_intptr_bool_System_Type_bool
        0x1026c47e0 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MacOS/MySingleView : ObjCRuntime_Runtime_GetINativeObject_T_REF_intptr_bool_bool
        0x1026c4768 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MacOS/MySingleView : ObjCRuntime_Runtime_GetINativeObject_T_REF_intptr_bool
        0x1026b6448 - ./xamarin-macios/tests/dotnet/MySingleView/bin/Release/net7.0-maccatalyst/maccatalyst-arm64/MySingleView.app/Contents/MacOS/MySingleView : Contacts_CNContactFormatter_GetDescriptorForRequiredKeys_Contacts_CNContactFormatterStyle
        ...
```

The comment is still outdated and I'll try to change the wording of the workaround description.

File: tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs [hunk 9]
```diff
@@ -281,6 +291,233 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
...
+						il.Emit (OpCodes.Call, abr.NativeObject_op_Implicit_IntPtr);
+					il.Emit (OpCodes.Newobj, ctor);
+					il.Emit (OpCodes.Ret);
+
+					il.Append (falseTarget);
+				}
+
+				// In addition to the big lookup method, implement the static factory method on the type:
+				ImplementConstructNSObjectFactoryMethod (type, ctor);
+			}
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs:336]: When I build `monotouchtest` in Debug, the code sizes of IL that I got from ILSpy in `Microsoft.MacCatalyst.dll` are:

| Function | Code size |
| :-- | --: |
| `__Registrar__.ConstructINativeObject` | `29288` |
| `__Registrar__.ConstructNSObject` | `96865` |
| `__Registrar__.LookupTypeId` | `121571` |
| `__Registrar__.LookupType` | `57911` |

File: src/ObjCRuntime/Runtime.cs (hunk 4 above)
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs:1316]: In this case I replaced the `class` constraint with `NSObject`. It's not possible to have both.

File: src/ObjCRuntime/INativeObject.cs (hunk 6 above)
Comment (simonrozsival) [reply src/ObjCRuntime/INativeObject.cs]: I made the interface method internal. That seems to be enough for the API comparison tool. The custom linker step changes it to public if MSR is used. I did the same goes for the `INSObjectFactory` interface and that seems to be OK as well.

File: tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs [hunk 10]
```diff
@@ -281,6 +291,251 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
...
+			Annotations.Mark (createInstanceMethod);
+		}
+
+		static MethodReference? FindNSObjectConstructor (TypeDefinition type)
+		{
+			return FindConstructorWithOneParameter ("ObjCRuntime", "NativeHandle")
+				?? FindConstructorWithOneParameter ("System", "IntPtr");
+
+			MethodReference? FindConstructorWithOneParameter (string ns, string cls)
+				=> type.Methods.FirstOrDefault (method =>
```
Comment (rolfbjarne) [inline tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs:515]: This seems safer, since in case the selection logic is wrong and matches more than one constructor:

```suggestion
				=> type.Methods.SingleOrDefault (method =>
```

File: tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs [hunk 11]
```diff
@@ -281,6 +291,251 @@ void GenerateLookupType (AssemblyTrampolineInfo infos, TypeDefinition registrarT
...
+		}
+
+
+		static MethodReference? FindINativeObjectConstructor (TypeDefinition type)
+		{
+			return FindConstructorWithTwoParameters ("ObjCRuntime", "NativeHandle", "System", "Boolean")
+				?? FindConstructorWithTwoParameters ("System", "IntPtr", "System", "Boolean");
+
+			MethodReference? FindConstructorWithTwoParameters (string ns1, string cls1, string ns2, string cls2)
+				=> type.Methods.FirstOrDefault (method =>
```
Comment (rolfbjarne) [inline tools/dotnet-linker/Steps/ManagedRegistrarLookupTablesStep.cs:530]:
```suggestion
				=> type.Methods.SingleOrDefault (method =>
```

File: src/ObjCRuntime/Runtime.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (rolfbjarne) [reply src/ObjCRuntime/Runtime.cs]: Did you file an issue for this? Seems like an issue in the AOT compiler.

File: docs/website/mtouch-errors.md [hunk 12]
```diff
@@ -3749,3 +3749,28 @@ This exception will have an inner exception which gives the reason for the failu
 ### MT8036: Failed to convert the value at index {index} from {type} to {type}.
 
 This exception will have an inner exception which gives the reason for the failure.
+
+<a name="MT8037" />
+
+### MT8037: Failed to marshal the Objective-C object {handle} (type: {objc_type}). Could not find an existing managed instance for this object, nor was it possible to create a new managed instance of generic type {managed_type}.
+
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply docs/website/mtouch-errors.md:3756]: I see, OK, I'll make those changes. Thanks!

File: src/ObjCRuntime/Runtime.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs]: I haven't done that yet but I will create a follow-up compiler issue.

File: src/ObjCRuntime/Runtime.cs [hunk 13]
```diff
@@ -1753,7 +1927,7 @@ static Type LookupINativeObjectImplementation (IntPtr ptr, Type target_type, Typ
 					// native objects and NSObject instances.
 					throw ErrorHelper.CreateError (8004, $"Cannot create an instance of {implementation.FullName} for the native object 0x{ptr:x} (of type '{Class.class_getName (Class.GetClassForObject (ptr))}'), because another instance already exists for this native object (of type {o.GetType ().FullName}).");
 				}
-				return ConstructNSObject<INativeObject> (ptr, implementation, MissingCtorResolution.ThrowConstructor1NotFound, sel, method_handle);
+				return (INativeObject?) ConstructNSObject<NSObject> (ptr, implementation!, MissingCtorResolution.ThrowConstructor1NotFound, sel, method_handle);
```
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs:1930]: You're right, it's not necessary anymore and I'll remove it.

File: src/ObjCRuntime/Runtime.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs]: https://github.com/dotnet/runtime/issues/90139

**Review by dalexsoto — COMMENTED** [review]
Looking good to me 👍

File: src/ObjCRuntime/Runtime.cs [hunk 14]
```diff
@@ -1310,19 +1337,61 @@ static void MissingCtor (IntPtr ptr, IntPtr klass, Type type, MissingCtorResolut
...
 			if (type is null)
 				throw new ArgumentNullException (nameof (type));
+#if NET
+			if (Runtime.IsManagedStaticRegistrar) {
+				T? instance = default;
+				var nativeHandle = new NativeHandle (ptr);
+
+				// We want to create an instance of `type` and if we have the chance to use the factory method
+				// on the generic type, we will prefer it to using the lookup table.
+				if (typeof (T) == type // TODO can I drop the IsGenericType check? would it break some test?
```
Comment (rolfbjarne) [inline src/ObjCRuntime/Runtime.cs:1362]: Could you resolve this TODO somehow?

**Review by rolfbjarne — APPROVED** [review]

File: src/ObjCRuntime/Runtime.cs (hunk 14 above)
Comment (simonrozsival) [reply src/ObjCRuntime/Runtime.cs:1362]: Thanks for pointing this out. I think the condition has changed since I wrote the comment and I just forgot to remove the comment. I'll double check that the comment is outdated and I'll get rid of it or address it.
